### PR TITLE
Add IAM permissions for each role

### PR DESCRIPTION
**Pull request by kylelaker-ccpo**
This creates three new roles intended to be used by humans:
1. `AtatQa` for QA/Testing
2. `AtatSecurityAuditor` for Security Auditors
3. `AtatDeveloper` for Developers

This also creates one new role to be passed by developers to CloudFormation
using the `--role-arn` or equivalent flag when building using the CDK:
1. `AtatCloudFormation`

Access is currently granted to the three job function roles to the
entire Central/Management account in AWS Organizations. Long-term, that
is not a good idea and we should find a way to lock it down a bit but
right now this gives us two significant benefits:

  1. Any user can assume any job role and see what those users see,
  which will be especially helpful for developers to help other roles
  troubleshoot other roles' IAM
  2. It simplifies the configuration and steps to assume the job-based
  roles compared to needing to double-hop

Most of the logic around assuming roles is going to need fairly
significant refactoring to support IAM Identity Federation in the near
future anyway (so that we are not relying on IAM Users directly). So
this feels like a pretty good interim solution but we can try to put in
the effort to lock it down if required.

The roles have the following privileges:

 - QA has access to view all "safe" resources, the items in DynamoDB,
   and to GET all API Gateway restapi resources
 - Auditors have all QA access plus access to all CloudWatch Logs, and
   all actions in the SecurityAudit AWS-managed job-function policy
 - Developers have all the access of QA plus read access to all
   CloudWatch Logs, as well as permission to use the CloudFormation
   execution role
 - The CloudFormation execution role has AdministratorAccess

All roles have an explicit Deny for all actions involving AWS
Organizations; however, this is not enforced by an SCP or Permissions
Boundary so we do allow new roles to be created with those permissions
(but it would have to be done with CloudFormation).

The Roles are all made available as CloudFormation stack exports. The
really cool thing about this is that it allows other stacks to grant the
roles access to specific resources that get created later (though there
are not any examples of that added in this PR since there is not really
a need).

We also make use of a pretty basic CloudFormation Custom Resource to get
the Account ID of the AWS Organizations Management Account. We are only
using one field from the output of that API call.

Finally, creation of the IAM stack at all is guarded by an environment
variable. Since we are creating named IAM resources, creating duplicate
stacks would result in failures and folks have gotten accustomed to
using `--all` on the CDK CLI. This adds a warning if the stack gets
created at all to remind users that everything will get wonky if we try
to spin up two if these in the same account (actually, it'll just
totally fail to create but making it sound spookier hopefully encourages
users to pay extra attention to whether it's intentional). The only
environment where multiple stacks being created is probable is the
Sandbox environment but it's also the most commonly-used environment.

So in summary there's a lot of room for improvement here:
 - using federated identity instead of IAM users in the central account
 - limiting the permissions on the CloudFormation role
 - moving to having resources grant access to specific ATAT roles
 - find some more explicit denies
 - improve the trust process

This gets us a good start, likely gives users the majority of access
that they need for day-to-day work, and is a solid base to continue
iterating on.

Ticket: AT-6563

### CODE REVIEW

File: infrastructure/lib/atat-iam-stack.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        awsManagedAuditorPolicy,
+        baseDenies,
+      ],
+    });
+    // Developers have all the read access of QA and Testing and the ability to pass
+    // the CloudFormation execution role and read all logs in CloudWatch.
+    const developerRole = new iam.Role(this, "DeveloperRole", {
+      roleName: "AtatDeveloper",
+      assumedBy: managementAccountPrincipal,
+      managedPolicies: [awsManagedViewOnlyPolicy, generalReadAccess, awsManagedLogsReadPolicy, baseDenies],
```
[the remark this answers is not in the document]
Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-iam-stack.ts]: Added CloudFormationFullAccess

File: infrastructure/lib/atat-iam-stack.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    });
+
+    const developerCdkStagingAccess = new iam.ManagedPolicy(this, "DeveloperCdkStagingAccess", {
+      statements: [
+        new iam.PolicyStatement({
+          sid: "AllowModifyingCdkToolBuckets",
+          effect: iam.Effect.ALLOW,
+          actions: ["s3:*"],
+          resources: [
+            `arn:${cdk.Aws.PARTITION}:s3:::cdk-hnb659fds-assets-${cdk.Aws.ACCOUNT_ID}-*`,
```
Comment (jeffsegal-ccpo) [inline infrastructure/lib/atat-iam-stack.ts:62]: Intentionally hard-coded ID here? Could you add a comment describing what it is?

Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:62]: yeah so that's a great point. from: https://docs.aws.amazon.com/cdk/latest/guide/bootstrapping.html

> `--qualifier` a string that is added to the names of all resources in the bootstrap stack. A qualifier lets you avoid name clashes when you provision two bootstrap stacks in the same environment. The default is `hnb659fds` (this value has no significance). Changing the qualifier will require changes to your AWS CDK app (see Stack synthesizers).

There might actually be a better way to get this value. I'll look into it.

File: infrastructure/lib/atat-iam-stack.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+          actions: ["dynamodb:*GetItem", "dynamodb:PartiQLSelect", "dynamodb:Scan", "dynamodb:Query"],
+          resources: [`arn:${cdk.Aws.PARTITION}:dynamodb:*:${cdk.Aws.ACCOUNT_ID}:table/*`],
+        }),
+        new iam.PolicyStatement({
+          sid: "APIGatewayRestApiReadAccess",
+          effect: iam.Effect.ALLOW,
+          actions: ["apigateway:GET"],
+          resources: [`arn:${cdk.Aws.PARTITION}:apigateway:*::/restapis*`],
+        }),
+      ],
```
[the remark this answers is not in the document]
Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:41]: `sqs:ListQueues` will be available through the `ViewOnlyPolicy`. Do we need to grant `sqs:*` or is there something more reasonable? Can we solve it in the SQS ticket?

File: infrastructure/lib/atat-iam-stack.ts (hunk 2 above)
Comment (jasonburkert-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:62]: Maybe including a comment with that URL is good enough.

**Review by jasonburkert-ccpo — APPROVED** [review]
Looks good assuming inclusion of a comment re: default qualifier value `hnb659fds`

File: infrastructure/lib/atat-iam-stack.ts (hunk 2 above)
Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:62]: Going to just change this to `cdk-*-assets` just in case we use other qualifiers in the future. And I actually will review Access Analyzer to see if this line of the policy ever gets used once it's in place. We may really only need the staging bucket.

File: infrastructure/lib/atat-iam-stack.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:41]: I guess the question is _who_ will need to perform those actions. Everyone, or developers? Perhaps these RW actions can go in the developer RW access policy? In any case, since SQS is not on the `develop` branch today and adding permissions for a service we're not currently using feels antithetical to the goals of trimming down IAM to what's required as part of this PR :) (ignore the CloudFormation role's permissions when you read that sentence). But I want to avoid speculatively granting access to things.

[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply infrastructure/lib/atat-iam-stack.ts:41]: I've asked @hannahposeyscholl-ccpo to create a future ticket referencing this conversation.

**Review by jeffsegal-ccpo — APPROVED** [review]

**Comment by kylelaker-ccpo** [comment]
Merge conflict has been resolved and no fundamental changes were made in the last push. Once the status checks come back, I'll merge this.

**Comment by kylelaker-ccpo** [comment]
Also need to add AWS Artifact access to the auditor role
